app/src/main/java/com/asfoundation/wallet/billing/mipay/MiPayViewModel.kt

```diff
@@ -0,0 +1,275 @@
+      Single.just(transactionBuilder)
+        .observeOn(rxSchedulers.io)
+        .doOnSuccess { transaction ->
+          analytics.sendPaymentErrorWithDetailsAndRiskEvent(
+            transaction.domain,
+            transaction.skuId,
+            transaction.amount().toString(),
+            BillingAnalytics.PAYMENT_METHOD_MI_PAY,
+            transaction.type,
+            errorCode ?: "",
+            errorMessage ?: "",
+            ""
+          )
+        }
+        .subscribe({}, { it.printStackTrace() })
+    )
+  }
+
+  private fun getTransactionStatus() {
+    transactionUid?.let { uid ->
+      getTransactionStatusUseCase(
+        uid = uid
+      ).map {
+        when (it.status) {
+          Transaction.Status.COMPLETED -> {
+            stopTransactionStatusTimer()
+            sendSideEffect {
+              successInfo = SuccessInfo(
+                hash = it.hash,
+                orderReference = null,
+                purchaseUid = it.uid,
+              )
+              MiPayIABSideEffect.ShowSuccess
+            }
+          }
+
+          Transaction.Status.INVALID_TRANSACTION,
+          Transaction.Status.FAILED,
+          Transaction.Status.CANCELED,
+          Transaction.Status.FRAUD -> {
+            stopTransactionStatusTimer()
+            sendSideEffect {
+              MiPayIABSideEffect.ShowError(
+                R.string.purchase_error_wallet_block_code_403
+              )
+            }
+          }
+
+          Transaction.Status.PENDING,
+          Transaction.Status.PENDING_SERVICE_AUTHORIZATION,
+          Transaction.Status.PROCESSING,
+          Transaction.Status.PENDING_USER_PAYMENT,
+          Transaction.Status.SETTLED -> {
+            sendSideEffect { MiPayIABSideEffect.ShowLoading }
+          }
+        }
+      }.scopedSubscribe()
+    }
+  }
+
+  fun getSuccessBundle(
+    transactionBuilder: TransactionBuilder?
+  ) {
+    if (transactionBuilder == null) {
+      sendSideEffect { MiPayIABSideEffect.ShowError(R.string.unknown_error) }
+      return
+    }
+    inAppPurchaseInteractor.savePreSelectedPaymentMethod(
+      PaymentMethodsView.PaymentMethodId.MI_PAY.id
+    )
+    createSuccessBundleUseCase(
+      transactionBuilder.type,
+      transactionBuilder.domain,
+      transactionBuilder.skuId,
+      successInfo?.purchaseUid,
+      successInfo?.orderReference,
+      successInfo?.hash,
+      networkScheduler
+    ).doOnSuccess {
+      sendSideEffect { MiPayIABSideEffect.SendSuccessBundle(it.bundle) }
+    }.subscribeOn(viewScheduler).observeOn(viewScheduler).doOnError {
+      sendSideEffect { MiPayIABSideEffect.ShowError(R.string.unknown_error) }
+    }.subscribe()
+  }
+
+  fun handleWebViewResult(result: ActivityResult) {
+    when (result.resultCode) {
+      WebViewActivity.SUCCESS -> {
+        startTransactionStatusTimer()
+      }
+
+      WebViewActivity.FAIL,
+      WebViewActivity.USER_CANCEL -> {
+        analytics.sendPaymentErrorEvent(
```

Comment:
On user cancelation we shouldn't send an error event, only if there was an actual error with the payment. On user cancel it should go back to the payment methods